### PR TITLE
[Pipelining] Fix PP grad scaling

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144734
* #144596
* __->__ #144352

Adds a grad-scaling method `perform_pp_grad_scaling()` which divides grads by num_microbatches.

Enables grad scaling by default, unless disabled due to using a loss function that sums instead of averaging losses.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/stage.py [hunk 1]
```diff
@@ -629,6 +637,7 @@ def perform_backward(
                     )
                 )
                 result = perform_backward(backward_type)()
+                perform_pp_grad_scaling()
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:640]: @fegin I think for DDP this won't work, becuase there is no way for me to insert pp_grad_scaling() call after backward and before DDP allreduce.

Actually, i'm not sure if this should matter.  I think it matters for FSDP becuase of how FSDP accumulates.

Comment (awgu) [reply torch/distributed/pipelining/stage.py:640]: Sorry, why does division before vs. after all-reduce/reduce-scatter matter?

Comment (wconstab) [reply torch/distributed/pipelining/stage.py:640]: I first experimentally found that doing scaling after running fsdp's post_backward did not work.  (it led to the PP test failing numerics).  I need to study the FSDP code.  I thought that FSDP was doing some accumulation trick and perhaps moving the data to a different buffer hence the order mattering.  Now that I write this, that doesn't make sense to me.  The p.grad field should be the source of truth both before and after the reduce_scatter, right?  (I was probably confusing myself with the case where fsdp accumulates into .main_grad, but that's not relevant here.)

Comment (awgu) [reply torch/distributed/pipelining/stage.py:640]: Aside from floating-point differences, I am not sure why doing the scaling after reduce-scatter would be an issue 🤔 
The `sharded_param.grad` should be the source of truth.

Is this test running with mixed precision? If not, then there is no separate tensor for accumulation. Even if so, the scaling is an in-place op, so I am not sure how there could be any issue here / I do not have any good sense of what could be happening :/

Comment (awgu) [reply torch/distributed/pipelining/stage.py:640]: > so, in pp stage code, wouldself.submod.parameters.foo would refer to an unsharded param or a sharded param? I think that's a trick question, the answer being "it depends whether you look at it before or after calling post_backward". Before post_backward, i'd be scaling the 'unsharded' parameter. After post_backward, i'd be scaling the sharded parameter. Is that right?

I agree that it depends on before or after the last post-backward, but what does not make sense is why before or after matters. Both should be mathematically equivalent.

Comment (wconstab) [reply torch/distributed/pipelining/stage.py:640]: yea. i think mathematically they should be equivalent.  I confirmed that if I apply some horrendous scaling like `p.grad /= 1000000`, it only seems to impact the test if I do it before post_backward.  If I do it after post_backward, the test behaves the same as if I do no scaling at all.  So I am not thinking about this as a numerics issue, but a mechanical one.  Probably somehow I'm modifying a wrong tensor.

File: torch/distributed/pipelining/stage.py [hunk 2]
```diff
@@ -639,6 +648,7 @@ def perform_backward(
             self.submod.set_requires_gradient_sync(False)
             result = perform_backward(backward_type)()
             if last_backward:
+                perform_pp_grad_scaling()
```
Comment (tianyu-l) [inline torch/distributed/pipelining/stage.py:651]: I hit the assertion error in `perform_pp_grad_scaling` if put before `run_post_backward`. It seemed to work after I switch the order.

File: torch/distributed/pipelining/stage.py [hunk 3]
```diff
@@ -616,6 +616,15 @@ def perform_backward(
             else:
                 raise RuntimeError(f"Unknown backward type: {backward_type}")
 
+        def perform_pp_grad_scaling():
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:619]: @H-Huang any thoughts on the overall way i plumbed this from schedule to stage?

i thought about how schedule is the place where loss fn is registered and also the place that tracks `last_backward`, so if we want to expose a user-API it might make sense for that to be in PipelineSchedule class.

Comment (H-Huang) [reply torch/distributed/pipelining/stage.py:619]: Yeah I agree the user-API should be at the `PipelineSchedule` class as well.

If we do the pp_grad_scaling after DP reduction, then the scaling can be entirely at the schedule implementation instead of passing in the scale factor into stage. Do you think that would be a better place? (though we would have to figure out the numerics differences between scaling before or after you mentioned below)

Comment (wconstab) [reply torch/distributed/pipelining/stage.py:623]: yea, it might be cleaner to move scaling into the schedule.  I was thinking at the time that it would require putting the scaling code in multiple schedule impls.  If we could do it cleanly, then i'd prefer that.

File: torch/distributed/pipelining/stage.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/stage.py]: ok. in conclusion, the before/after thing was not relevant for numerics per-se.  I even did a simulation in the reference model where I scaled by 8 first and 2 later, vs 2 first and 8 later, and as expected this made no difference.  The only issues in the end were
 - PP/FSDP sync bug (prev PR https://github.com/pytorch/pytorch/pull/144535) made it so scaling after was racy
 - mixed-precision stores grads differently, making it so scaling before is more complicated (and leads to  missing p.grad errors if done naively)
 
So after fixing the race issue, the scaling implementation is to always scale after, which is also simple for mixed-precision.

File: torch/distributed/pipelining/stage.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/pipelining/stage.py]: scaling is now done after reduction, so this issue should be solved.  mixed-precision is covered by the correctness test now.

**Review by H-Huang — APPROVED** [review]
Looks good to me. We discussed offline about reading the `.reduction` attribute from `loss_fn` to determine the scale factor, so that can be implemented in a follow up PR

File: torch/distributed/pipelining/stage.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/pipelining/stage.py]: i made this change. scaling is now called from the schedule.

**Comment by wconstab** [comment]
> We discussed offline about reading the .reduction attribute from loss_fn to determine the scale factor, so that can be implemented in a follow up PR

I think in the end what I want to do is this:
* accept `loss_fn` as `Union[Callable, _Loss]`
* add `scale_grads: bool = None` kwarg
* if users provide loss_fn of type `_Loss` we will infer scale_grads behavior from the loss fn unless `scale_grads` is also specified, which overrides it
* otherwise, we default to what's specified in `scale_grads`, if specified
* finally, if neither loss_fn nor scale_grads specifies anything, we use 'True'

My bias is more to be transparent and discoverable and make it convenient regardless of how the loss_fn is implemented.

I think I'll work on landing this PR as-is and do that in a followup, unless you don't like keeping the `scale_grads` kwarg.

**Comment by H-Huang** [comment]
1. accept loss_fn as Union[Callable, _Loss]
2. add scale_grads: bool = None kwarg
3. if users provide loss_fn of type _Loss we will infer scale_grads behavior from the loss fn unless scale_grads is also specified, which overrides it
4. otherwise, we default to what's specified in scale_grads, if specified
5. finally, if neither loss_fn nor scale_grads specifies anything, we use 'True'

This makes sense to me. For the scale grads args it should probably default to `scale_grads: bool = True` if we are setting it to `True` in step 5 as a default, that will be less confusing from explaining the `None` to `True`. Then for the loss_fn.reduction that will be used to determine the value it is scaled by (or no scaling if scale_grads is set to `False`).

**Comment by wconstab** [comment]
> This makes sense to me. For the scale grads args it should probably default to scale_grads: bool = True if we are setting it to True in step 5 as a default, that will be less confusing from explaining the None to True. Then for the loss_fn.reduction that will be used to determine the value it is scaled by (or no scaling if scale_grads is set to False).

the issue is that if `scale_grads` defaults to `True` then it always overrides whatever comes from loss_fn.  

I think it's clearer if we document that =None means it will default to True if not specified via loss_fn, and then `=True` or `=False`  is a clear user override.
